### PR TITLE
Teardown all internal components on exception

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

When an exception is raised, we were only tearing down the data connector. But we should tear all internal components.

This is required for a follow-up where the data fetchers are moved to the loops.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -685,14 +685,13 @@ def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs:
...
             self.state.status = TrainerStatus.INTERRUPTED
             if distributed_available() and self.world_size > 1:
-                # try syncing remaing processes, kill otherwise
+                # try syncing remaining processes, kill otherwise
                 self.strategy.reconciliate_processes(traceback.format_exc())
             self._on_exception()
+            self._call_callback_hooks("on_exception", exception)
+            self._teardown()
             # reset bookkeeping
             self.state.stage = None
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:694]: The stage needs to be the last thing to unset as `teardown` might access it

**Review by rohitgr7 — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (justusschock) [reply pytorch_lightning/trainer/trainer.py:694]: Maybe we should add a comment for that to prevent that from changing again?

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -685,14 +685,13 @@ def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs:
         except BaseException as exception:
             self.state.status = TrainerStatus.INTERRUPTED
             if distributed_available() and self.world_size > 1:
-                # try syncing remaing processes, kill otherwise
+                # try syncing remaining processes, kill otherwise
                 self.strategy.reconciliate_processes(traceback.format_exc())
             self._on_exception()
+            self._call_callback_hooks("on_exception", exception)
+            self._teardown()
             # reset bookkeeping
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:693]:
```suggestion
            # teardown might access the stage so we reset it after
```

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -1240,7 +1239,10 @@ def _teardown(self):
         self.strategy.post_dispatch(self)
         self.strategy.teardown()
         self._data_connector.teardown()
-        self._active_loop.teardown()
+        loop = self._active_loop
+        # loop should never be `None` here but it can because we don't know the trainer stage with `ddp_spawn`
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1243]: makes sense. note, we recover the trainer stage in
`                self.strategy._recover_results_in_main_process(spawn_output, self)
`
L673 in trainer.py

**Review by awaelchli — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

**Pull request merged**